File: src/gr6j/outputs.py
"""Structured output dataclasses for GR6J model results.

This module provides dataclasses for organizing and accessing model outputs:
- GR6JOutput: GR6J model flux outputs
- SnowOutput: CemaNeige snow module flux outputs
- SnowLayerOutputs: Per-layer snow outputs for multi-layer simulations
- ModelOutput: Combined model output with time index
"""

from __future__ import annotations

from dataclasses import dataclass, fields

import numpy as np
import pandas as pd


@dataclass(frozen=True)
class GR6JOutput:
    """GR6J model flux outputs as arrays.

    All arrays have the same length as the input forcing data.

    Attributes:
        pet: Potential evapotranspiration [mm/day].
        precip: Precipitation input to GR6J [mm/day]. When snow module is
            enabled, this is the liquid water output from CemaNeige.
        production_store: Production store level after timestep [mm].
        net_rainfall: Net rainfall after interception [mm/day].
        storage_infiltration: Water infiltrating to production store [mm/day].
        actual_et: Actual evapotranspiration [mm/day].
        percolation: Percolation from production store [mm/day].
        effective_rainfall: Total effective rainfall after percolation [mm/day].
        q9: Output from UH1 (90% branch) [mm/day].
        q1: Output from UH2 (10% branch) [mm/day].
        routing_store: Routing store level after timestep [mm].
        exchange: Groundwater exchange potential [mm/day].
        actual_exchange_routing: Actual exchange from routing store [mm/day].
        actual_exchange_direct: Actual exchange from direct branch [mm/day].
        actual_exchange_total: Total actual exchange [mm/day].
        qr: Outflow from routing store [mm/day].
        qrexp: Outflow from exponential store [mm/day].
        exponential_store: Exponential store level after timestep [mm].
        qd: Direct branch outflow [mm/day].
        streamflow: Total simulated streamflow [mm/day].
    """

    pet: np.ndarray
    precip: np.ndarray
    production_store: np.ndarray
    net_rainfall: np.ndarray
    storage_infiltration: np.ndarray
    actual_et: np.ndarray
    percolation: np.ndarray
    effective_rainfall: np.ndarray
    q9: np.ndarray
    q1: np.ndarray
    routing_store: np.ndarray
    exchange: np.ndarray
    actual_exchange_routing: np.ndarray
    actual_exchange_direct: np.ndarray
    actual_exchange_total: np.ndarray
    qr: np.ndarray
    qrexp: np.ndarray
    exponential_store: np.ndarray
    qd: np.ndarray
    streamflow: np.ndarray

    def to_dict(self) -> dict[str, np.ndarray]:
        """Convert to dictionary of arrays.

        Returns:
            Dictionary mapping field names to their numpy array values.
        """
        return {field.name: getattr(self, field.name) for field in fields(self)}


@dataclass(frozen=True)
class SnowOutput:
    """CemaNeige snow module flux outputs as arrays.

    All arrays have the same length as the input forcing data.

    Attributes:
        precip_raw: Original precipitation before snow processing [mm/day].
        snow_pliq: Liquid precipitation (rain) [mm/day].
        snow_psol: Solid precipitation (snow) [mm/day].
        snow_pack: Snow pack water equivalent after melt [mm].
        snow_thermal_state: Thermal state of snow pack [deg C].
        snow_gratio: Snow cover fraction after melt [-].
        snow_pot_melt: Potential melt [mm/day].
        snow_melt: Actual melt [mm/day].
        snow_pliq_and_melt: Total liquid output to GR6J [mm/day].
        snow_temp: Air temperature [deg C].
        snow_gthreshold: Melt threshold [mm].
        snow_glocalmax: Local maximum for hysteresis [mm].
    """

    precip_raw: np.ndarray
    snow_pliq: np.ndarray
    snow_psol: np.ndarray
    snow_pack: np.ndarray
    snow_thermal_state: np.ndarray
    snow_gratio: np.ndarray
    snow_pot_melt: np.ndarray
    snow_melt: np.ndarray
    snow_pliq_and_melt: np.ndarray
    snow_temp: np.ndarray
    snow_gthreshold: np.ndarray
    snow_glocalmax: np.ndarray

    def to_dict(self) -> dict[str, np.ndarray]:
        """Convert to dictionary of arrays.

        Returns:
            Dictionary mapping field names to their numpy array values.
        """
        return {field.name: getattr(self, field.name) for field in fields(self)}


@dataclass(frozen=True)
class SnowLayerOutputs:
    """Per-layer snow outputs for multi-layer CemaNeige simulations.

    Contains 2D arrays with shape (n_timesteps, n_layers) for detailed
    layer-by-layer analysis. The catchment-aggregated values are still
    available in SnowOutput.

    Attributes:
        layer_elevations: Representative elevation of each layer [m]. Shape (n_layers,).
        layer_fractions: Area fraction of each layer [-]. Shape (n_layers,).
        snow_pack: Snow pack water equivalent per layer [mm]. Shape (n_timesteps, n_layers).
        snow_thermal_state: Thermal state per layer [deg C]. Shape (n_timesteps, n_layers).
        snow_gratio: Snow cover fraction per layer [-]. Shape (n_timesteps, n_layers).
        snow_melt: Actual melt per layer [mm/day]. Shape (n_timesteps, n_layers).
        snow_pliq_and_melt: Total liquid output per layer [mm/day]. Shape (n_timesteps, n_layers).
        layer_temp: Extrapolated temperature per layer [deg C]. Shape (n_timesteps, n_layers).
        layer_precip: Extrapolated precipitation per layer [mm/day]. Shape (n_timesteps, n_layers).
    """

    layer_elevations: np.ndarray  # Shape (n_layers,)
    layer_fractions: np.ndarray  # Shape (n_layers,)
    snow_pack: np.ndarray  # Shape (n_timesteps, n_layers)
    snow_thermal_state: np.ndarray  # Shape (n_timesteps, n_layers)
    snow_gratio: np.ndarray  # Shape (n_timesteps, n_layers)
    snow_melt: np.ndarray  # Shape (n_timesteps, n_layers)
    snow_pliq_and_melt: np.ndarray  # Shape (n_timesteps, n_layers)
    layer_temp: np.ndarray  # Shape (n_timesteps, n_layers)
    layer_precip: np.ndarray  # Shape (n_timesteps, n_layers)

    @property
    def n_layers(self) -> int:
        """Return the number of elevation layers."""
        return len(self.layer_elevations)

    def to_dict(self) -> dict[str, np.ndarray]:
        """Convert to dictionary of arrays.

        Returns:
            Dictionary mapping field names to their numpy array values.
        """
        return {field.name: getattr(self, field.name) for field in fields(self)}


@dataclass(frozen=True)
class ModelOutput:
    """Complete model output combining GR6J and optional snow outputs.

    Provides a unified interface for accessing all model results with time
    indexing and conversion to pandas DataFrame.

    Attributes:
        time: Datetime array for each timestep.
        gr6j: GR6J model outputs.
        snow: Optional CemaNeige outputs (present if snow module was enabled).
        snow_layers: Optional per-layer snow outputs for multi-layer simulations.
    """

    time: np.ndarray
    gr6j: GR6JOutput
    snow: SnowOutput | None = None
    snow_layers: SnowLayerOutputs | None = None

    def __len__(self) -> int:
        """Return the number of timesteps."""
        return len(self.time)

    def to_dataframe(self) -> pd.DataFrame:
        """Convert to DataFrame with time index.

        Combines all GR6J outputs (and snow outputs if present) into a single
        DataFrame with time as the index.

        Returns:
            DataFrame with all flux outputs and time as index.
        """
        data = self.gr6j.to_dict()

        if self.snow is not None:
            data.update(self.snow.to_dict())

        df = pd.DataFrame(data, index=self.time)
        df.index.name = "time"

        return df

File: src/gr6j/outputs.rs
//! GR6J model flux outputs.
//!
//! Two levels: `Fluxes` holds a single timestep, `FluxesTimeseries` holds
//! the full simulation (Vec of each field). Mirrors the airGR MISC array (1-20).

/// Single-timestep fluxes — returned by `step()`.
#[derive(Debug, Clone, Copy)]
pub struct Fluxes {
    pub pet: f64,
    pub precip: f64,
    pub production_store: f64,
    pub net_rainfall: f64,
    pub storage_infiltration: f64,
    pub actual_et: f64,
    pub percolation: f64,
    pub effective_rainfall: f64,
    pub q9: f64,
    pub q1: f64,
    pub routing_store: f64,
    pub exchange: f64,
    pub actual_exchange_routing: f64,
    pub actual_exchange_direct: f64,
    pub actual_exchange_total: f64,
    pub qr: f64,
    pub qrexp: f64,
    pub exponential_store: f64,
    pub qd: f64,
    pub streamflow: f64,
}

/// Full timeseries of fluxes — returned by `run()`.
#[derive(Debug)]
pub struct FluxesTimeseries {
    pub pet: Vec<f64>,
    pub precip: Vec<f64>,
    pub production_store: Vec<f64>,
    pub net_rainfall: Vec<f64>,
    pub storage_infiltration: Vec<f64>,
    pub actual_et: Vec<f64>,
    pub percolation: Vec<f64>,
    pub effective_rainfall: Vec<f64>,
    pub q9: Vec<f64>,
    pub q1: Vec<f64>,
    pub routing_store: Vec<f64>,
    pub exchange: Vec<f64>,
    pub actual_exchange_routing: Vec<f64>,
    pub actual_exchange_direct: Vec<f64>,
    pub actual_exchange_total: Vec<f64>,
    pub qr: Vec<f64>,
    pub qrexp: Vec<f64>,
    pub exponential_store: Vec<f64>,
    pub qd: Vec<f64>,
    pub streamflow: Vec<f64>,
}

impl FluxesTimeseries {
    /// Pre-allocate all vectors for `n` timesteps.
    pub fn with_capacity(n: usize) -> Self {
        Self {
            pet: Vec::with_capacity(n),
            precip: Vec::with_capacity(n),
            production_store: Vec::with_capacity(n),
            net_rainfall: Vec::with_capacity(n),
            storage_infiltration: Vec::with_capacity(n),
            actual_et: Vec::with_capacity(n),
            percolation: Vec::with_capacity(n),
            effective_rainfall: Vec::with_capacity(n),
            q9: Vec::with_capacity(n),
            q1: Vec::with_capacity(n),
            routing_store: Vec::with_capacity(n),
            exchange: Vec::with_capacity(n),
            actual_exchange_routing: Vec::with_capacity(n),
            actual_exchange_direct: Vec::with_capacity(n),
            actual_exchange_total: Vec::with_capacity(n),
            qr: Vec::with_capacity(n),
            qrexp: Vec::with_capacity(n),
            exponential_store: Vec::with_capacity(n),
            qd: Vec::with_capacity(n),
            streamflow: Vec::with_capacity(n),
        }
    }

    /// Push a single timestep's fluxes into the timeseries.
    pub fn push(&mut self, f: &Fluxes) {
        self.pet.push(f.pet);
        self.precip.push(f.precip);
        self.production_store.push(f.production_store);
        self.net_rainfall.push(f.net_rainfall);
        self.storage_infiltration.push(f.storage_infiltration);
        self.actual_et.push(f.actual_et);
        self.percolation.push(f.percolation);
        self.effective_rainfall.push(f.effective_rainfall);
        self.q9.push(f.q9);
        self.q1.push(f.q1);
        self.routing_store.push(f.routing_store);
        self.exchange.push(f.exchange);
        self.actual_exchange_routing.push(f.actual_exchange_routing);
        self.actual_exchange_direct.push(f.actual_exchange_direct);
        self.actual_exchange_total.push(f.actual_exchange_total);
        self.qr.push(f.qr);
        self.qrexp.push(f.qrexp);
        self.exponential_store.push(f.exponential_store);
        self.qd.push(f.qd);
        self.streamflow.push(f.streamflow);
    }

    /// Number of timesteps.
    pub fn len(&self) -> usize {
        self.streamflow.len()
    }

    /// Returns `true` if there are no timesteps.
    pub fn is_empty(&self) -> bool {
        self.streamflow.is_empty()
    }
}
